File: go/internal/prop/prop.go
// Package prop provides Android system property read and write.
package prop

import (
	"fmt"
	"strings"

	"github.com/Limplom/nothingctl/internal/adb"
	nterrors "github.com/Limplom/nothingctl/internal/errors"
)

var prefixGroups = []string{
	"ro.product",
	"ro.build",
	"ro.boot",
	"ro.hardware",
	"persist",
	"sys",
	"gsm",
	"net",
	"wifi",
}

func groupKey(propName string) string {
	for _, prefix := range prefixGroups {
		if strings.HasPrefix(propName, prefix) {
			return prefix
		}
	}
	return "other"
}

func parseGetprop(output string) [][2]string {
	var props [][2]string
	for _, line := range strings.Split(output, "\n") {
		line = strings.TrimRight(line, "\r")
		if !strings.HasPrefix(line, "[") {
			continue
		}
		idx := strings.Index(line, "]:")
		if idx == -1 {
			continue
		}
		key := line[1:idx]
		rest := strings.TrimSpace(line[idx+2:])
		value := rest
		if strings.HasPrefix(rest, "[") && strings.HasSuffix(rest, "]") {
			value = rest[1 : len(rest)-1]
		}
		if key != "" {
			props = append(props, [2]string{key, value})
		}
	}
	return props
}

// ActionPropGet reads one or all system properties.
// key="" means dump all grouped by prefix.
func ActionPropGet(serial, model, key string) error {
	if key != "" {
		out, _, _ := adb.Run([]string{"adb", "-s", serial, "shell",
			fmt.Sprintf("getprop %s", key)})
		val := strings.TrimSpace(out)
		if val == "" {
			fmt.Printf("  [WARN] Property '%s' is empty or not set.\n", key)
		} else {
			fmt.Printf("  %s = %s\n", key, val)
		}
		return nil
	}

	stdout, stderr, code := adb.Run([]string{"adb", "-s", serial, "shell", "getprop"})
	if code != 0 {
		return nterrors.AdbError(fmt.Sprintf("getprop failed: %s", strings.TrimSpace(stderr)))
	}

	props := parseGetprop(stdout)
	if len(props) == 0 {
		return nterrors.AdbError("getprop returned no output.")
	}

	// Group by prefix
	grouped := make(map[string][][2]string)
	for _, g := range prefixGroups {
		grouped[g] = nil
	}
	grouped["other"] = nil

	for _, kv := range props {
		g := groupKey(kv[0])
		grouped[g] = append(grouped[g], kv)
	}

	fmt.Printf("\n  System Properties \u2014 Nothing %s\n\n", model)

	groupOrder := append(prefixGroups, "other")
	for _, g := range groupOrder {
		entries := grouped[g]
		if len(entries) == 0 {
			continue
		}
		fmt.Printf("  [%s.*]\n", g)
		maxLen := 0
		for _, kv := range entries {
			if len(kv[0]) > maxLen {
				maxLen = len(kv[0])
			}
		}
		for _, kv := range entries {
			fmt.Printf("    %-*s  = %s\n", maxLen, kv[0], kv[1])
		}
		fmt.Println()
	}
	return nil
}

// ActionPropSet writes a system property via root su.
// Requires Magisk root.
func ActionPropSet(serial, key, value string) error {
	if !adb.CheckAdbRoot(serial) {
		return nterrors.AdbError(
			"Root not available via ADB shell.\n" +
				"Enable in Magisk: Settings -> Superuser access -> Apps and ADB.")
	}

	fmt.Println("  NOTE: Most ro.* properties reset on reboot. Use persist.* for persistent changes.")

	if strings.HasPrefix(key, "ro.") {
		fmt.Println("  [WARN] ro.* properties are read-only at the system level \u2014 this may not persist.")
	}

	_, stderr, code := adb.Run([]string{"adb", "-s", serial, "shell",
		fmt.Sprintf(`su -c "setprop %s %s"`, key, value)})
	if code != 0 {
		errMsg := strings.TrimSpace(stderr)
		return nterrors.AdbError(fmt.Sprintf("setprop %s failed: %s", key, errMsg))
	}

	fmt.Printf("  [OK] Set %s = %s\n", key, value)
	return nil
}

File: python/nothing/prop.py
"""Read and write Android system properties via ADB."""

from .backup import check_adb_root
from .device import adb_shell, run
from .exceptions import AdbError
from .models import DeviceInfo

# ---------------------------------------------------------------------------
# Property grouping
# ---------------------------------------------------------------------------

# Ordered list of prefix groups; first match wins.
_PREFIX_GROUPS = [
    "ro.product",
    "ro.build",
    "ro.boot",
    "ro.hardware",
    "persist",
    "sys",
    "gsm",
    "net",
    "wifi",
]


def _group_key(prop_name: str) -> str:
    """Return the display-group name for a property key."""
    for prefix in _PREFIX_GROUPS:
        if prop_name.startswith(prefix):
            return prefix
    return "other"


# ---------------------------------------------------------------------------
# Parsing helpers
# ---------------------------------------------------------------------------

def _parse_getprop(output: str) -> list[tuple[str, str]]:
    """
    Parse `adb shell getprop` output.

    Lines are formatted as:  [key]: [value]
    Returns a list of (key, value) tuples preserving declaration order.
    """
    props: list[tuple[str, str]] = []
    for line in output.splitlines():
        line = line.strip()
        if not line.startswith("["):
            continue
        # Format: [key]: [value]
        bracket_close = line.find("]:")
        if bracket_close == -1:
            continue
        key = line[1:bracket_close]
        rest = line[bracket_close + 2:].strip()
        # Value is wrapped in [ … ]
        if rest.startswith("[") and rest.endswith("]"):
            value = rest[1:-1]
        else:
            value = rest
        if key:
            props.append((key, value))
    return props


# ---------------------------------------------------------------------------
# action_prop_get
# ---------------------------------------------------------------------------

def action_prop_get(device: DeviceInfo, key: str | None) -> None:
    """
    Read system properties.

    If *key* is given, print its value directly.
    If *key* is None, dump all properties grouped by prefix.
    """
    if key:
        value = adb_shell(f"getprop {key}", device.serial)
        if not value:
            print(f"  [WARN] Property '{key}' is empty or not set.")
        else:
            print(f"  {key} = {value}")
        return

    # Full dump
    r = run(["adb", "-s", device.serial, "shell", "getprop"])
    if r.returncode != 0:
        raise AdbError(f"getprop failed: {r.stderr.strip()}")

    props = _parse_getprop(r.stdout)
    if not props:
        raise AdbError("getprop returned no output.")

    # Bucket into groups, preserving order within each group
    grouped: dict[str, list[tuple[str, str]]] = {g: [] for g in _PREFIX_GROUPS}
    grouped["other"] = []

    for key_name, value in props:
        group = _group_key(key_name)
        grouped[group].append((key_name, value))

    print(f"\n  System Properties — Nothing {device.model}\n")

    group_order = _PREFIX_GROUPS + ["other"]
    for group in group_order:
        entries = grouped.get(group, [])
        if not entries:
            continue

        print(f"  [{group}.*]")
        # Align values: find longest key in this group
        max_len = max(len(k) for k, _ in entries)
        for k, v in entries:
            print(f"    {k:<{max_len}}  = {v}")
        print()


# ---------------------------------------------------------------------------
# action_prop_set
# ---------------------------------------------------------------------------

def action_prop_set(device: DeviceInfo, key: str, value: str) -> None:
    """
    Write a system property via root su.

    Requires Magisk root. Most ro.* properties reset on reboot; use persist.*
    for persistent changes.
    """
    if not check_adb_root(device.serial):
        raise AdbError(
            "Root not available via ADB shell.\n"
            "Enable in Magisk: Settings -> Superuser access -> Apps and ADB."
        )

    print(
        "  NOTE: Most ro.* properties reset on reboot. "
        "Use persist.* for persistent changes."
    )

    if key.startswith("ro."):
        print(
            f"  [WARN] ro.* properties are read-only at the system level "
            f"— this may not persist."
        )

    r = run(["adb", "-s", device.serial, "shell", f'su -c "setprop {key} {value}"'])
    if r.returncode != 0:
        stderr = r.stderr.strip() or r.stdout.strip()
        raise AdbError(f"setprop {key} failed: {stderr}")

    print(f"  [OK] Set {key} = {value}")
